File: sparkcql-core/src/main/scala/gov/va/sparkcql/adapter/model/ModelAdapter.scala
package gov.va.sparkcql.adapter.model

import scala.reflect.runtime.universe._
import scala.collection.JavaConverters._
import javax.xml.namespace.QName
import org.apache.spark.sql.types.StructType
import gov.va.sparkcql.io.Log

trait ModelAdapter {

  val namespaceUri: String

  val supportedDataTypes: List[QName]

  def assertDataTypeIsSupported(dataType: QName): Unit = {
    if (!supportedDataTypes.contains(dataType)) {
      throw new RuntimeException("Unsupported data type '" + dataType.toString() + "'.")
    }
  }

  def schemaOf(dataType: QName): Option[StructType]
  
  def metaInterval(typeName: String): (String, String)

  def typeToElmMapping(typeName: String): Map[String, String]

  def patientIdentifier(): QName = ???    // TODO: Need a way to identify a patient and relating b/w models
}

File: sparkcql-core/src/main/java/gov/va/sparkcql/adapter/model/ModelAdapter.java
package gov.va.sparkcql.adapter.model;

import java.util.List;

import javax.xml.namespace.QName;

import org.apache.spark.sql.types.StructType;

import gov.va.sparkcql.adapter.Adapter;

public interface ModelAdapter extends Adapter {
    
    public String namespaceUri();

    public List<QName> supportedDataTypes();
    
    public StructType schemaOf(QName dataType);
    
    default void assertDataTypeSupport(QName dataType) {
        if (!supportedDataTypes().contains(dataType)) {
            throw new RuntimeException("Unsupported data type '" + dataType.toString() + "'.");
        }
    }

    // public QName patientIdentifier();
}